types/policy.go

```diff
@@ -170,151 +170,136 @@ func (p SpendPolicy) String() string {
 }
 
 // ParseSpendPolicy parses a spend policy from a string.
-func ParseSpendPolicy(p string) (SpendPolicy, error) {
-	parenBegin := strings.IndexByte(p, '(')
-	if parenBegin == -1 {
-		return SpendPolicy{}, errors.New("policy missing begin parentheses")
+func ParseSpendPolicy(s string) (SpendPolicy, error) {
+	sp, rem, err := parseSpendPolicy(s)
+	if err != nil {
+		return SpendPolicy{}, err
+	} else if rem != "" {
+		return SpendPolicy{}, fmt.Errorf("trailing bytes: %q", rem)
 	}
-	parenEnd := strings.LastIndexByte(p, ')')
-	if parenEnd == -1 {
-		return SpendPolicy{}, errors.New("policy missing end parentheses")
-	}
-	if parenBegin > parenEnd {
-		return SpendPolicy{}, errors.New("close parentheses before open parentheses")
+	return sp, nil
+}
+
+func parseSpendPolicy(s string) (SpendPolicy, string, error) {
+	var sp SpendPolicy
+	nextToken := func() string {
+		s = strings.TrimSpace(s)
+		i := strings.IndexAny(s, "(),[]")
+		if i == -1 {
+			return ""
+		}
+		t := s[:i]
+		s = s[i:]
+		return t
 	}
+	consume := func(b byte) error {
```

Comment:
consider using a "sticky error" here so that you don't have to check the return value of `consume` at each callsite